src/backends/gpu_kernels.hpp

```diff
@@ -0,0 +1,434 @@
+        const std::vector<T>& in,
+        const std::vector<I>& sizes,
+        const std::vector<I>& starts,
+        int block_width, int num_mtx, int padded_length)
+{
+    auto num_blocks = impl::block_count(num_mtx, block_width);
+    std::vector<T> out(num_blocks*block_width*padded_length, impl::npos<T>());
+    for (auto mtx: util::make_span(0, num_mtx)) {
+        auto block = mtx/block_width;
+        auto lane  = mtx%block_width;
+
+        auto len = sizes[mtx];
+        auto src = starts[mtx];
+        auto dst = block*(block_width*padded_length) + lane;
+        for (auto i: util::make_span(0, len)) {
+            out[dst] = in[src+i];
+            dst += block_width;
+        }
+    }
+    return out;
+};
+
+template <typename T, typename I, int BlockWidth, int LoadWidth, int THREADS>
+__global__
+void reverse_interleave(
+    const T* in, T* out, const I* sizes, const I* starts, int padded_size, int num_mtx)
+{
+    static_assert(BlockWidth*LoadWidth==THREADS, "");
+
+    __shared__ T buffer[THREADS];
+
+    const auto tid = threadIdx.x + blockIdx.x*blockDim.x;
+    const auto lid = threadIdx.x;
+
+    const auto mtx_id   = tid/LoadWidth;
+    const auto mtx_lane = tid - mtx_id*LoadWidth;
+
+    const auto blk_id   = tid/(BlockWidth*LoadWidth);
+    const auto blk_row  = lid/BlockWidth;
+    const auto blk_lane = lid - blk_row*BlockWidth;
+
+    const auto blk_pos  = LoadWidth*blk_lane + blk_row;
+
+    const bool do_store = mtx_id<num_mtx;
+
+    // only threads that participate in storing access starts and sizes arrays
+    auto store_pos = do_store? starts[mtx_id] + mtx_lane     : 0;
+    const auto end = do_store? starts[mtx_id] + sizes[mtx_id]: 0;
+    auto load_pos  = blk_id*BlockWidth*padded_size + (blk_row*BlockWidth + blk_lane);
+
+    for (auto i=0; i<padded_size; i+=LoadWidth) {
+        auto loaded = impl::npos<T>();
+        if (i+blk_row<padded_size) {
+            loaded = in[load_pos];
+        }
+        buffer[blk_pos] = loaded;
+        __syncthreads();
+        if (do_store && store_pos<end) {
+            out[store_pos] = buffer[lid];
+        }
+        load_pos  += LoadWidth*BlockWidth;
+        store_pos += LoadWidth;
+    }
+}
+
+template <typename T, typename I, int BlockWidth, int LoadWidth>
+void reverse_interleave(const T* in, T* out, const I* sizes, const I* starts, int padded_size, int num_mtx)
+{
+    constexpr int Threads = BlockWidth*LoadWidth;
+    const int blocks = impl::block_count(num_mtx, BlockWidth);
+
+    reverse_interleave<T, I, BlockWidth, LoadWidth, Threads>
+        <<<blocks, Threads>>> (in, out, sizes, starts, padded_size, num_mtx);
+}
+
+/// GPU implementatin of Hines matrix assembly
+/// Flat layout
+/// For a given time step size dt
+///     - use the precomputed alpha and alpha_d values to construct the diagonal
+///       and off diagonal of the symmetric Hines matrix.
+///     - compute the RHS of the linear system to solve
+template <typename T, typename I>
```

Comment:
It is a floating point type. This is consistent throughout pretty much the whole of NestMC. So it is consistent.

That isn't to say that it is necessarily a great naming scheme, however if we rename, we should do so everywhere for consistency. Which is a topic for another PR.